src/Microsoft.AspNetCore.Mvc.Modules/ModularApplicationPart.cs

```diff
@@ -74,18 +75,23 @@ public IEnumerable<TypeInfo> Types
         {
             get
             {
-                return DefaultAssemblyDiscoveryProvider
-                    .DiscoverAssemblies(
-                        ShellBlueprint.Dependencies.Select(dep => dep.Type.GetTypeInfo().Assembly).Distinct().ToList(),
-                        ReferenceAssemblies)
+                var assemblies = ModularAssemblyProvider
+                    .GetAssemblies(ExtensionLibraryService.RuntimeLibraries);
+
+                var types = DefaultModularAssemblyDiscoveryProvider
+                    .GetCandidateAssemblies(assemblies, ReferenceAssemblies)
                     .SelectMany(x => x.DefinedTypes);
+
+                return types;
             }
         }
 
         /// <inheritdoc />
         public IEnumerable<string> GetReferencePaths()
         {
-            return AssemblyProvider.GetAssemblies().Select(x => x.Location);
+            return DependencyContext.Default.CompileLibraries
```

Comment:
I think the modules should have preservecompilationcontext on them.... then we get the compilation libraries from them as they are what we are compiling.